xc/device.ml

```diff
@@ -1963,6 +1974,96 @@ module Dm = struct
+          Epoll.ctl epfd Epoll.EPOLL_CTL_DEL (Qmp_protocol.to_fd c) [Epoll.EPOLLIN];
+          Qmp_protocol.close c
+        with
+        | Unix.Unix_error(Unix.EBADF, "close", p) -> raise(Internal_error (Printf.sprintf "Failed to close qmp socket on domain-%d: %s" domid p))
+        | e -> raise(Internal_error (Printf.sprintf "Get error trying remove qmp on domain-%d: %s" domid (Printexc.to_string e)))
+      end
+    in
+
+    let add domid =
+      if not(Hashtbl.mem domid_to_connection_tbl domid)
+      then begin
+        let path = Printf.sprintf "/var/run/xen/qmp-event-libxl-%d" domid in
+        try
+          let c = Qmp_protocol.connect path in
+          Qmp_protocol.negotiate c;
+          Hashtbl.replace domid_to_connection_tbl domid c;
+          Hashtbl.replace connection_to_domid_tbl c domid;
+          Epoll.ctl epfd Epoll.EPOLL_CTL_ADD (Qmp_protocol.to_fd c) [Epoll.EPOLLIN];
+          debug "QMP domain-%d: negotiation complete" domid
+        with e ->
+          debug "QMP domain-%d: negotiation failed (%s): removing socket" domid (Printexc.to_string e);
+          remove domid
+      end
+    in
+
+    while true do
+      let cs = Hashtbl.fold (fun k v acc -> v :: acc) domid_to_connection_tbl [] in
+      let fds = List.map (fun x -> Qmp_protocol.to_fd x, x) cs in
+      let rs = Epoll.wait epfd (Hashtbl.length domid_to_connection_tbl + 1) (-1) in
+      List.iter (
+        fun ef ->
+          let events = fst ef in
+          let fd = snd ef in
+          if fd = qmp_event_pipe_r then begin
+            List.iter (
+              function
+              | Epoll.EPOLLIN ->
+                add (int_of_string (input_line qmp_event_channel_r))
+              | Epoll.EPOLLRDHUP | Epoll.EPOLLHUP | Epoll.EPOLLERR ->
+                raise(Internal_error "Epoll error in qmp_event_thread")
+              | _ -> ()
+            ) events
+          end else begin
+            let c = List.assoc fd fds in
+            let domid = Hashtbl.find connection_to_domid_tbl c in
+            List.iter (
+              function
+              | Epoll.EPOLLIN ->
+                  if Hashtbl.mem domid_to_connection_tbl domid
+                  then begin
+                    try
+                      let m = Qmp_protocol.read c in
+                      match m with 
+                      | Event e -> qmp_event_handle domid e
+                      | _ -> debug "Get non-event message, domain-%d: %s" domid (string_of_message m)
+                    with End_of_file ->
+                      debug "domain-%d: end of file, close qmp socket" domid;
+                      remove domid
+                  end
+              | Epoll.EPOLLRDHUP | Epoll.EPOLLHUP | Epoll.EPOLLERR -> 
+                  debug "EPOLL error on domain-%d, close qmp socket" domid;
+                  remove domid
+              | _ -> ()
+            ) events
+          end
+      ) rs;
+    done
+
+  let init_qmp_event () =
```

Comment:
no need for a function, just remove () and it will become a value that the ocaml runtime will execute when xenopsd starts.

Reply:
Hi @mg12 , we cannot let it executes automatically before `init()` in `xenops_server_xen.ml`, 
because if we run `xenopsd-xc  --help`, the thread also executes and the whole process will not return.
The strace log:
```
epoll_create1(EPOLL_CLOEXEC)            = 7
epoll_ctl(7, EPOLL_CTL_ADD, 3, {EPOLLIN, {u32=3, u64=3}}) = 0
epoll_wait(7, 
```